specs/networking/libp2p-standardization.md

```diff
@@ -0,0 +1,131 @@
+ETH 2.0 Networking Spec - Libp2p standard protocols
+===
+
+# Abstract
+
+Ethereum 2.0 clients plan to use the libp2p protocol networking stack for
+mainnet release. This document aims to standardize the libp2p client protocols,
+configuration and messaging formats.
+
+# Libp2p Protocols
+
+## Gossipsub
+
+#### Protocol id: `/meshsub/1.0.0`
+
+*Note: Parameters listed here are subject to a large-scale network feasibility
+study*
+
+The [Gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub)
+protocol will be used for block and attestation propagation across the
+network.
+
+### Configuration Parameters
+
+Gossipsub has a number of internal configuration parameters which directly
+effect the network performance.  Clients can implement independently, however
+we aim to standardize these across clients to optimize the gossip network for
+propagation times and message duplication. Current network-related defaults are:
+
+```
+(
+	// The target number of peers in the overlay mesh network (D in the libp2p specs).
+	mesh_size: 6
+	// The minimum number of peers in the mesh network before adding more (D_lo in the libp2p specs).
+	mesh_lo: 4
+	// The maximum number of peers in the mesh network before removing some (D_high in the libp2p sepcs).
+	mesh_high: 12
+	// The number of peers to gossip to during a heartbeat (D_lazy in the libp2p sepcs).
+	gossip_lazy: 6 // defaults to `mesh_size`
+	// Time to live for fanout peers (seconds).
+	fanout_ttl: 60
+	// The number of heartbeats to gossip about.
+	gossip_history: 3
+	// Time between each heartbeat (seconds).
+	heartbeat_interval: 1
+)
+```
+
+### Topics
+
+*The Go and Js implementations use string topics - This is likely to be
+updated to topic hashes in later versions - https://github.com/libp2p/rust-libp2p/issues/473*
+
+For Eth2.0 clients, topics will be sent as `SHA2-256` hashes of the topic string.
+
+There is one dedicated topic for propagating beacon blocks and aggregated
+attestations across the network. This topic will have the string
+`beacon_chain`. Each shard will have it's own topic allowing relevant parties
+to subscribe to in order to receive local shard attestations. The shard topics are
+prefixed with `shard` followed by the number of the shard. For example,
+messages relating to shard 10, will have the topic string `shard10`.
+
+### Messages
+
+Messages sent across gossipsub are fixed-size length-prefixed byte arrays.
+Each message has a maximum size of 512KB (estimated from expected largest uncompressed
+block size).
+
+The byte array is prefixed with a unsigned 64 bit length number encoded as an
```

Comment:
since ssz is binary little endian uint64 (8 bytes), why not follow suit here? fewer "necessary" encodings in the app layer

Reply:
Yep, I also don't like the idea of having extra encodings. The reason this is here, is because this is currently how `rust-libp2p` sends raw messages over floodsub/gossipsub. I have explicitly written it out for other implementers not using libp2p. 
I assume this is the same for the other libp2p implementations. This particular encoding wasn't a design choice, rather highlighting what currently exists. 
This being said, I'm happy to modify libp2p to suit our needs.

Reply:
This exists in libp2p to prevent arbitrary sized byte streams. Once a message is received, libp2p checks the initial size. If the initial size is lager than the maximum, it drops the message. It then proceeds to read `min(max_size, size_of_message)` bytes. 
We could modify this to deal only SSZ encoding, but I believe it's changing the actual floodsub/gossipsub spec (assuming the spec is the implementations :) ).

Reply:
so fwiw, my comment was primarily aimed at keeping things natural for each layer - if the length is part of "libp2p" and not the "eth2 payload", it should of course use whatever libp2p tends to use. echoing @jannikluhn 's sentiment here - should this be part of an eth2 document, or just noted as "assume the transport is a block transport / provides length"?

Reply:
Yep. I added it here because I was concerned about client interop and wanted to specifically lay out the bytes across the wire. I agree that it doesn't belong in this document and clients can sort it out if their libp2p implementations don't match. Will remove this.